### PR TITLE
[golang] Update to 1.19rc2 (from 1.18.3)

### PR DESCRIPTION
**Pull request by dims**
Go 1.19 GA release is planned for the first week of August:
https://groups.google.com/g/golang-dev/c/4xsD_D5oflI/m/77Ofgt_EAAAJ?utm_medium=email&utm_source=footer

With this PR, we would update to go1.19 rc2 and when the GA is available, we would move to that.

Signed-off-by: Davanum Srinivas <davanum@gmail.com>

#### What type of PR is this?
/kind feature

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
For v1.25, Kubernetes will be using golang 1.19, In this PR we update to 1.19rc2 as GA is not yet available.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by liggitt** [comment]
huh... did 1.19 gofmt make all those comment changes?

**Comment by liggitt** [comment]
answering my own question: yes - https://tip.golang.org/doc/go1.19#go-doc

**Comment by dims** [comment]
@liggitt ack thanks. I added a commit after running `hack/update-gofmt.sh` with 1.19rc2

**Comment by liggitt** [comment]
> @liggitt ack thanks. I added a commit after running `hack/update-gofmt.sh` with 1.19rc2

If previous versions of gofmt leave comments alone, I think we could hoist the result of running 1.19rc2 `gofmt` into a standalone PR and go ahead and merge it ahead of bumping 1.18 → 1.19 if we want to, to keep this one as small as possible

**Comment by dims** [comment]
> > @liggitt ack thanks. I added a commit after running `hack/update-gofmt.sh` with 1.19rc2
> 
> If previous versions of gofmt leave comments alone, I think we could hoist the result of running 1.19rc2 `gofmt` into a standalone PR and go ahead and merge it ahead of bumping 1.18 → 1.19 if we want to, to keep this one as small as possible

good point! will do so

**Comment by aojea** [comment]
are we try to get this in 1.25?

**Comment by dims** [comment]
> are we trying to get this in 1.25?

Dunno yet. I am exploring how much work is needed for sure. 

- https://kubernetes.slack.com/archives/C2C40FMNF/p1658173936650579
- https://kubernetes.slack.com/archives/CHGFYJVAN/p1658159216005969

**Comment by dims** [comment]
NOTE: see that `pull-kubernetes-verify` is failing but `pull-kubernetes-verify-go-canary` is passing, this is because the verify job is still using go1.18 while the verify-go-canary was bumped to golang 1.19rc2 in https://github.com/kubernetes/test-infra/pull/26860/files

**Comment by dims** [comment]
(simple rebase, no changes!)

File: pkg/fieldpath/fieldpath.go
Comment (liggitt) [inline pkg/fieldpath/fieldpath.go:94]: this is actually wrong... did gofmt make this change?

File: pkg/kubelet/pluginmanager/cache/types.go
Comment (liggitt) [inline pkg/kubelet/pluginmanager/cache/types.go:36]: oof... gofmt is not kind to ascii art

File: staging/src/k8s.io/kubectl/pkg/util/fieldpath/fieldpath.go
Comment (liggitt) [inline staging/src/k8s.io/kubectl/pkg/util/fieldpath/fieldpath.go:96]: same here, this syntax is wrong now

**Comment by liggitt** [comment]
do we have capability to run a canary integration job on 1.19rc2? I remember the bump to 1.18 hit integration issues once CI moved

File: pkg/fieldpath/fieldpath.go
Comment (dims) [reply pkg/fieldpath/fieldpath.go:94]: yes, it did.

Let me replace that example using double quotes `metadata.labels[”]"`

File: staging/src/k8s.io/kubectl/pkg/util/fieldpath/fieldpath.go
Comment (dims) [reply staging/src/k8s.io/kubectl/pkg/util/fieldpath/fieldpath.go:96]: same ... switching to double quotes

File: pkg/fieldpath/fieldpath.go
Comment (liggitt) [reply pkg/fieldpath/fieldpath.go:94]: but double quotes aren't valid syntax for fieldPath indexing... only single-quotes are :-/

Comment (liggitt) [reply pkg/fieldpath/fieldpath.go:94]: is there a way to indent or wrap this so gofmt leaves it alone?

Comment (liggitt) [reply pkg/fieldpath/fieldpath.go:94]: on second thought, this might actually be a bug worth reporting to go... indenting and tweaking list bullets is one thing... replacing characters inside godoc strings is much more invasive

Comment (liggitt) [reply pkg/fieldpath/fieldpath.go:94]: looks like we want these to be code blocks:

> A code block is a span of indented or blank lines not starting with a bullet list marker or numbered list marker. It is rendered as preformatted text (a <pre> block in HTML).

maybe dropping the leading list markers and putting blank lines between them would make gofmt leave these alone?

File: pkg/kubelet/pluginmanager/cache/types.go
Comment (dims) [reply pkg/kubelet/pluginmanager/cache/types.go:36]: fixing. i am not quite sure how this got tripped up.

**Comment by dims** [comment]
@liggitt @thockin this is ready to merge. Please take a look.

**Comment by dims** [comment]
also see https://github.com/kubernetes/kubernetes/issues/111396

File: staging/src/k8s.io/kubectl/pkg/cmd/cmd.go
Comment (liggitt) [inline staging/src/k8s.io/kubectl/pkg/cmd/cmd.go:177]: Since 1.19 gives us a specific error indicator for the dot-path lookup, can we use that rather than string checks?

I think that would limit the changes in this file to:
```diff
- if err != nil || len(path) == 0 {
+ if shouldSkipOnLookPathErr(err) || len(path) == 0 {
```

then the <=go1.18 impl is:
```go
// +build !go1.19
...
func shouldSkipOnLookPathErr(err error) bool {
  return err != nil
}
```

and the 1.19+ impl is:
```go
// +build go1.19
...
func shouldSkipOnLookPathErr(err error) bool {
  return err != nil && !errors.Is(err, exec.ErrDot)
}
```

File: staging/src/k8s.io/kubectl/pkg/cmd/cmd.go
Comment (liggitt) [inline staging/src/k8s.io/kubectl/pkg/cmd/cmd.go:196]: could use the same error checker helper here just to make sure we're not getting some other error we actually want to surface:

```go
lp, err := exec.LookPath(name)
if lp != "" && !shouldSkipOnLookPathErr(err) {
  cmd.Path = lp
```

**Comment by dims** [comment]
@MadhavJivrajani @liggitt please see the very last commit in this one, has the changes for Lookup/osexec stuff

File: staging/src/k8s.io/kubectl/pkg/cmd/cmd.go
Comment (liggitt) [inline staging/src/k8s.io/kubectl/pkg/cmd/cmd.go:196]: Were you able to try this on a windows machine (or ask someone from sig-windows to)? Do we also need to replace `Args[0]` with the fully resolved binary? (I don't know enough about how Command works on Windows to know)

**Comment by liggitt** [comment]
> @MadhavJivrajani @liggitt please see the very last commit in this one, has the changes for Lookup/osexec stuff

Thanks, all the changes I understand lgtm, had one question about the Args[0] use of the passed-in name

File: staging/src/k8s.io/kubectl/pkg/cmd/cmd.go
Comment (dims) [reply staging/src/k8s.io/kubectl/pkg/cmd/cmd.go:196]: @liggitt i am fairly sure this is fine as it is from https://github.com/golang/go/blob/14abe8aa7324bdf0e09e1dfebfb3519cc30f4918/src/os/exec/exec.go#L273-L278 :)

Comment (liggitt) [reply staging/src/k8s.io/kubectl/pkg/cmd/cmd.go:196]: ha, ok

File: hack/lib/golang.sh
Comment (liggitt) [inline hack/lib/golang.sh:485]: once this merges and CI switches to 1.19rc2, will people with a local go1.18 be generating/formatting changes that will fail CI?

Comment (dims) [reply hack/lib/golang.sh:485]: Yes, there is one file that has trouble, this is what happens when i run update-gofmt.sh using golang 1.18 (with this branch)

```
diff --git staging/src/k8s.io/apiserver/pkg/util/flowcontrol/patch_test.go staging/src/k8s.io/apiserver/pkg/util/flowcontrol/patch_test.go
index a5b01802cb95..b468b6d3b0d1 100644
--- staging/src/k8s.io/apiserver/pkg/util/flowcontrol/patch_test.go
+++ staging/src/k8s.io/apiserver/pkg/util/flowcontrol/patch_test.go
@@ -59,10 +59,10 @@ func Test_configController_generatePatchBytes(t *testing.T) {
                {
                        name: "check when message has a whitespace character that can be escaped",
                        condition: flowcontrol.FlowSchemaCondition{
-                               Type:               flowcontrol.FlowSchemaConditionDangling,
-                               Status:             flowcontrol.ConditionTrue,
-                               Reason:             "test reason",
-                               Message:            `test               none`,
+                               Type:   flowcontrol.FlowSchemaConditionDangling,
+                               Status: flowcontrol.ConditionTrue,
+                               Reason: "test reason",
+                               Message: `test          none`,
                                LastTransitionTime: metav1.NewTime(now),
                        },
                        want: []byte(fmt.Sprintf(`{"status":{"conditions":[{"type":"Dangling","status":"True","lastTransitionTime":"%s","reason":"test reason","message":"test \t\tnone"}]}}`, now.Format(time.RFC3339))),

```

The `Message` has a horizontal tab (hex 0x09) in the string and how that is handled is different between the go versions

**Comment by dims** [comment]
@liggitt ack thanks, will reorder stuff later today!

File: hack/lib/golang.sh
Comment (liggitt) [reply hack/lib/golang.sh:485]: interesting... 1.19 gofmt and 1.18 gofmt dueling seems like a bug we should report... my understanding is that gofmt tries to only make changes a previous version would leave alone

as a workaround for us, you could put a blank line before and after the `Message` line and I bet both go1.19 and go1.18 would format it identically

Comment (liggitt) [reply hack/lib/golang.sh:485]: > interesting... 1.19 gofmt and 1.18 gofmt dueling seems like a bug we should report

well, looks intentional, given https://github.com/golang/go/issues/51910 / https://github.com/golang/go/commit/bf68170c638e7e69bedcc64fadfd83354fd06c10

Comment (liggitt) [reply hack/lib/golang.sh:485]: the workaround to avoid adding formatting friction to ~everyone's PRs the week of code freeze is probably worthwhile

Comment (dims) [reply hack/lib/golang.sh:485]: `blank line before and after the Message line` did not help @liggitt 

Can i switch this to "test  \u0009none" (2 spaces and then a tab are in between `"test"` and `"one"`

Comment (liggitt) [reply hack/lib/golang.sh:485]: > Can i switch this to "test \u0009none" (2 spaces and then a tab are in between "test" and "one"

yup, that'd work too

Comment (dims) [reply hack/lib/golang.sh:485]: Done!

**Comment by dims** [comment]
@liggitt @thockin reordered/squashed commits as requested in https://github.com/kubernetes/kubernetes/pull/111254#issuecomment-1194207441

will ping when this CI goes green. (used diff to make sure i did not make a mistake in the reorder/squash process. rebased to master as well)

**Comment by liggitt** [comment]
Changes lgtm (and fwiw, I checked out the PR and ran gofmt with go1.18 and there were no changes, which is great)

Sweeping what 1.19 gofmt "autocorrected" in our doc strings, there's a fair number of places we had weird indents that gofmt is turning into new paragraphs, etc. I think this PR should go ahead and merge so we can get go1.19 signal, but a followup to sweep the doc auto format changes in this PR and fix up ones that look wrong would be good.

/lgtm
/approve
/hold to make sure canary presubmits are green

mechanically, once canary presubmits are green, I guess we would:
1. update real unit/integration/verify presubmits to use the go1.19rc2 images (which will make all other PRs to master start failing on format/verify errors)
2. retest this PR to get it to pass blocking presubmits and merge it

A heads-up that this is coming is probably worth communicating to dev@kubernetes.io ... the format changes on generated files is unfortunate, as it will make a ton of PRs require rebase

**Review by mrunalp — APPROVED** [review]

**Comment by liggitt** [comment]
pull-kubernetes-verify-go-canary passed

looks like all the verify failures were formatting ones which will resolve when we bump test-infra

**Comment by dims** [comment]
@liggitt yep that landed in https://github.com/kubernetes/test-infra/pull/26920 and i am retesting the 2 jobs after that.

**Comment by lavalamp** [comment]
This is causing an annoying gofmt change: https://github.com/kubernetes-sigs/structured-merge-diff/pull/220/files#diff-054f947623025b6f29ed3f11f2849794e09e83d9a31927598d4c1f059e7c4fe1

cc @alexzielenski @apelisse

**Comment by liggitt** [comment]
xref https://tip.golang.org/doc/go1.19#go-doc - As part of this change, [gofmt](https://tip.golang.org/cmd/gofmt) now reformats doc comments to make their rendered meaning clearer.

I'd recommend committing the go1.19 version of the doc, which go1.18 gofmt will leave alone

**Comment by apelisse** [comment]
Yeah, nested lists aren't supported in comments by gofmt which is unfortunate and clearly changes the meaning of some comments.

**Comment by alexzielenski** [comment]
filed an issue with golang https://github.com/golang/go/issues/54129

**Comment by lavalamp** [comment]
There's an example at the very bottom here: https://tip.golang.org/doc/comment

I guess they will probably close the issue :(

**Pull request merged**
